### PR TITLE
[17.0][MIG] webservice: Migration to 17.0

### PR DESCRIPTION
**Pull request by SilvioC2C**
Based on #34, includes #35 and #47 forwardport

### CODE REVIEW

**Comment by john-herholz-dt** [comment]
Thanks for taking over. Are you able to fix the tests? Should be not so complicated

File: webservice/tests/common.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            r.status_code = 200
+            return r
+        return super()._request_handler(s, r, **kw)
+
+
+@contextmanager
+def mock_cursor(cr):
+    with mock.patch("odoo.sql_db.Connection.cursor") as mocked_cursor_call:
+        org_close = cr.close
+        org_autocommit = cr.autocommit
```
Comment (imlopes) [inline webservice/tests/common.py]:
```suggestion
        org_autocommit = cr._cnx.autocommit
```

File: webservice/tests/common.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    def setUpClass(cls):
+        super().setUpClass()
+        cls._setup_env()
+        cls._setup_records()
+
+    @classmethod
+    def _request_handler(cls, s: Session, r: PreparedRequest, /, **kw):
+        if r.url.startswith("http://localhost.demo.odoo/"):
+            r = Response()
+            r.status_code = 200
```
Comment (imlopes) [inline webservice/tests/common.py]:
```suggestion
            r.status_code = 200
            r._content = b'{}'
```

**Comment by imlopes** [comment]
IMHO some errors are linked to https://github.com/OCA/OCB/commit/4111f02fb3cc4cdf3dd94adae0ca80cfe874170c
I've tried to fix, no success

**Comment by john-herholz-dt** [comment]
Hi there, I updated my PR for this as well but was struggling with the tests.
To mock the requests is really tricky. I stopped at the point in the tests where the responses object was checked.
When the request is mocked, the responses.calls were empty.

If anybody knows how to do this, please let us know.

```
 responses.add(responses.GET, self.url, body="{}")
 result = self.webservice.call("get")
```

Here the responses.calls will be empty

**Comment by john-herholz-dt** [comment]
@SilvioC2C congrats for the passed tests. Finally.
Have you been able to run tests locally. I always struggle to run OCA tests locally. Can you describe me how to set it up?
Just a demo database?

**Comment by SilvioC2C** [comment]
@john-herholz-dt @yankinmax @imlopes @gurneyalex tests are finally green, would you mind reviewing please? :pray:

**Review by john-herholz-dt — APPROVED** [review]

**Comment by SilvioC2C** [comment]
> Can you describe me how to set it up? Just a demo database?

Sorry, just saw your comment.
In C2C, we use a docker image that already contains test utils: https://github.com/camptocamp/docker-odoo-project#running-tests
Usually, I just run these commands:
```
docker-compose run --rm --name=odoo -e DB_NAME=testdb odoo testdb-gen -i <module_name>  # this creates a new DB and install the module I want to test
docker-compose run --rm --name=odoo -e DEMO=True -e DB_NAME=testdb -e MIGRATE=False odoo odoo --workers=0 --test-enable --stop-after-init -u <module_name>  # this updates the module and runs its tests
```

**Review by imlopes — APPROVED** [review]
Thanks for this huge work!!!

**Review by yankinmax — APPROVED** [review]
I see where I was wrong, thanks @SilvioC2C very nice

**Comment by john-herholz-dt** [comment]
@simahawk can you merge it please? I saw you merged the migration to 16.0.
Thank you!

**Review by simahawk — CHANGES_REQUESTED** [review]
Sorry, cannot be merged like this...
Issues:
* commits are a bit messed up or incomplete
* the mocking of the response seems useless as you should use `responses`

@SilvioC2C pls ping me and we can have a look together.

**Comment by ivantodorovich** [comment]
This PR includes:
- https://github.com/OCA/web-api/pull/47

However, that's not merged yet and by the looks of the reviewer's comments there's still some work to do there.
Could we take care of those comments and get approvals before attempting to forward port them here?

Alternatively, I'd suggest keeping the forward-port on a separate PR, so we can move on with the actual migration

**Comment by SilvioC2C** [comment]
@simahawk @ivantodorovich @imlopes @yankinmax @john-herholz-dt 

I have reworked the PR to take care of the pending PR #47 (which has been superseded by #63, now merged) and the tests (but I'm still open to comments and suggestions).
I could use an helping hand about the commit history, though.

In the meanwhile, would you mind giving this PR another round of reviews? :sweat_smile:

**Review by john-herholz-dt — APPROVED** [review]

**Review by yankinmax — APPROVED** [review]
Thanks @SilvioC2C , I like your changes regarding tests.

**Review by ivantodorovich — APPROVED** [review]

**Review by simahawk — APPROVED** [review]

**Pull request merged**